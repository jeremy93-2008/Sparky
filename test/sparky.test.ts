const lib = require("../src/sparky");
const Sparky = lib.Sparky;

describe("Diff method", () => {
    test("Nothing to render", () => {
        const diff = Sparky.reconciliate(null, null);
        expect(diff).toBe(null);
    });
    test("Only oldNode provided", () => {
        const diff = Sparky.reconciliate(lib.render(`<div>Hola a todos</div>`).dom, null);
        expect(diff).toBe(null);
    });

    test("Only newNode provided", () => {
        const currentDom = lib.render(`<div>Hola a todos</div>`).dom;
        const diff = Sparky.reconciliate(null, currentDom);
        expect(diff).toBe(currentDom);

        expect(diff.innerHTML).toBe(lib.render(`<div>Hola a todos</div>`).dom.innerHTML);
    });

    test("If same nodes are provided the result must be the same", () => {
        const currentDom = lib.render(`<div>Hola a todos</div>`).dom;
        const diff = Sparky.reconciliate(currentDom, lib.render(`<div>Hola a todos</div>`).dom);

        expect(diff).toBe(currentDom);
        expect(diff.innerHTML).toBe(lib.render(`<div>Hola a todos</div>`).dom.innerHTML);
    });

    test("Add a new node by diff", () => {
        const currentDom = lib.render(`<div>Hola a todos</div>`).dom;
        const diff = Sparky.reconciliate(currentDom, lib.render(`<div>Hola a todos <span>Jeremy</span></div>`).dom);

        expect(diff).toBe(currentDom);
        expect(diff.innerHTML).toBe(lib.render(`<div>Hola a todos <span>Jeremy</span></div>`).dom.innerHTML);
    });

    test("Replace a existing node by diff", () => {
        const currentDom = lib.render(`<div>Hola a todos <b>Before</b></div>`).dom;
        const diff = Sparky.reconciliate(currentDom, lib.render(`<div>Hola a todos <em>After</em></div>`).dom);

        expect(diff).toBe(currentDom);
        expect(diff.innerHTML).toBe(lib.render(`<div>Hola a todos <em>After</em></div>`).dom.innerHTML);
    });

    test("Adding a new nodes as a list by diff", () => {
        const currentDom = lib.render(`<div>Hola a todos <ul></ul></div>`).dom;
        const diff = Sparky.reconciliate(currentDom, lib.render(`<div>Hola a todos <ul>
            <li>Hola</li>
            <li>Adios</li>
            <li>Buenas</li>
        </ul></div>`).dom);

        expect(diff).toBe(currentDom);
        expect(diff.innerHTML).toBe(lib.render(`<div>Hola a todos <ul>
            <li>Hola</li>
            <li>Adios</li>
            <li>Buenas</li>
        </ul></div>`).dom.innerHTML);
    });

    test("Add new attributes to an element", () => {
        const currentDom = lib.render(`<div>Hola a todos <b>Before</b></div>`).dom;
        const diff = Sparky.reconciliate(currentDom, lib.render(`<div>Hola a todos <b class='selected'>After</b></div>`).dom);

        expect(diff).toBe(currentDom);
        expect(diff.innerHTML).toBe(lib.render(`<div>Hola a todos <b class='selected'>After</b></div>`).dom.innerHTML);
    });

    test("Add more attributes to an element", () => {
        const currentDom = lib.render(`<div>Hola a todos <b class='selected'>Before</b></div>`).dom;
        const diff = Sparky.reconciliate(currentDom, lib.render(`<div>Hola a todos <b id='uno' class='selected more'>After</b></div>`).dom);

        expect(diff).toBe(currentDom);
        expect(diff.innerHTML).toBe(lib.render(`<div>Hola a todos <b class='selected more' id='uno'>After</b></div>`).dom.innerHTML);
    });

    test("Add less attributes to an element", () => {
        const currentDom = lib.render(`<div>Hola a todos <b id='uno' class='selected more'>Before</b></div>`).dom;
        const diff = Sparky.reconciliate(currentDom, lib.render(`<div>Hola a todos <b class='selected'>After</b></div>`).dom);

        expect(diff).toBe(currentDom);
        expect(diff.innerHTML).toBe(lib.render(`<div>Hola a todos <b class='selected'>After</b></div>`).dom.innerHTML);
    });

    test("Change attributes to an element", () => {
        const currentDom = lib.render(`<div>Hola a todos <b class='selected more'>Before</b></div>`).dom;
        const diff = Sparky.reconciliate(currentDom, lib.render(`<div>Hola a todos <b class='selected'>After</b></div>`).dom);

        expect(diff).toBe(currentDom);
        expect(diff.innerHTML).toBe(lib.render(`<div>Hola a todos <b class='selected'>After</b></div>`).dom.innerHTML);
    });
})

describe("Mount function", () => {
    test("Basic function", () => {
        Sparky.mount(Sparky.component(() => {
            return lib.render /*html*/`
            <div>
                <p>Hola a todos</p>
            </div>
            `
        }), document.body);
        expect(document.body).toMatchSnapshot()
    });
    test("Computed function", () => {
        Sparky.mount(Sparky.component(() => {
            const a = 24;
            return lib.render /*html*/`
            <div>
                <p>Hola a todos ${a}</p>
            </div>
            `
        }), document.body);
        // console.log(document.body.innerHTML)
        expect(document.body).toMatchSnapshot()
    });
    test("Computed w/ State function", () => {
        Sparky.mount(Sparky.component((self) => {
            const a = self.getState("boton") || "uno";
            return lib.render /*html*/`
            <div>
                <p>Hola a todos ${a}</p>
            </div>
            `
        }), document.body);
        // console.log(document.body.innerHTML)
        expect(document.body).toMatchSnapshot()
    });
    test("Computed, State w/ Event function", () => {
        Sparky.mount(Sparky.component((self) => {
            const a = self.getState("boton") || "uno";
            const onClick = () => {
                self.setState({"boton" : 10})
            }
            return lib.render /*html*/`
            <div>
                <p onclick=${onClick}>Hola a todos ${a}</p>
            </div>
            `
        }), document.body);
        // console.log(document.body.innerHTML)
        expect(document.body).toMatchSnapshot()
    });
    test("Computed, State, Memo w/ Event and Click function", () => {

        const Span = (self) => {
            const a = self.getState("boton") || "uno";
            return lib.render /*html*/`
            <div>
                <p>Hola a todos ${a}</p>
            </div>
            `
        };

        Sparky.mount(Sparky.component((self) => {
            if(!self.getState("boton")) self.setState({boton: "hola mundo"})
            const a = self.getState("boton") || "uno";
            
            self.memoize(() => {
                console.log("Memo!")
            }, [a]);

            self.onUpdate(() => {
                console.log("Update!")
            }, [a])

            const onClick = () => {
                self.setState({"boton" : 10})
            }
            
            return lib.render /*html*/`
            <div>
                <p id="paragraph" onclick=${onClick}>Hola a todos ${a}</p>
                <div>${Sparky.component(Span)}</div>
            </div>
            `
        }), document.body);
        // console.log(document.body.innerHTML)
        expect(document.body).toMatchSnapshot()
    });
})