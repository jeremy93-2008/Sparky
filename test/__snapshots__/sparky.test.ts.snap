// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mount function Basic function 1`] = `
<body>
  <div>
    
                
    <p>
      Hola a todos
    </p>
    
            
  </div>
</body>
`;

exports[`Mount function Computed function 1`] = `
<body>
  <div>
    
                
    <p>
      Hola a todos 
      <span
        class="computed"
      >
        24
      </span>
    </p>
    
            
  </div>
</body>
`;

exports[`Mount function Computed w/ State function 1`] = `
<body>
  <div>
    
                
    <p>
      Hola a todos 
      <span
        class="computed"
      >
        uno
      </span>
    </p>
    
            
  </div>
</body>
`;
